### ISSUE TITLE
TLS Trusted CA Indication extension

### ISSUE THREAD

**Issue by securitykernel** [randombit/botan#1183]
We are currently planning to add support for the [Trusted CA Indication](https://tools.ietf.org/html/rfc6066#page-12) to Botan. I think the following things would need to be done:

Client side:
* Need to add extension encoding to `TLS::Extension`, client hello
* `Credentials_Manager` interface already has function `trusted_ca_certificates()`, which returns the CAs a TLS client trusts
* Clients do not always want to send the CA list by default for privacy and/or efficiency reasons, so add a function `bool trusted_ca_indication()` to the `TLS_Policy` class, which defaults to `false`

Server side
* Incorporate trusted_ca_indication information into routine selecting the server certificate shown to the client

For compatibility tests, I only found MatrixSSL to support this extension, neither OpenSSL nor GnuTLS do.

Open questions:
* Client can send either key hash, cert hash or DN; do we decide for one of these or do we make it configurable for applications?
* Does the TLS server select the appropriate server certificate alone or do we want to pass the Trusted CA information to the application via `Credentials_Manager::cert_chain_single_type()`?

**Comment by randombit** [comment]
OK with me.

Re what format to send in, it seems like supporting all 3 is a bit overmuch. Beyond that I guess I don't have a firm opinion here. Supporting only one form seems like it would lead to the simplest implementation so if possible let's do that. The key hash format has a special case for RSA which I would rather avoid, so I'd lean (weakly) for either DN or cert hash. I think a big part of this consideration may be also what other implementations support. BTW on that front  I looked at the MatrixSSL source (https://github.com/matrixssl/matrixssl) and didn't find anything relevant to this extension, what did I miss?

Oh, also if we support just `DN`, the same `Credentials_Manager` API could be used for handling DNs sent in  `CertificateRequest`messages which would be nice. So I guess my preferred ordering would be 1) DN 2) cert hash 3) key hash. And TBH I don't see any real privacy issues for sending DNs that don't also occur for sending unkeyed hashes.

If 2 forms must be supported, for configuration maybe the answer is having `trusted_ca_indication` return an `enum class Trusted_CA_Indicator_Type` with values `None`, `Cert_Hash`, `DN`.

Re selection, it seems best to pass this information along to the `Credentials_Manager`. In doing so we may want to take the opportunity to add a new API that allows providing multiple relevant details which currently are not passed from the TLS server to the cert selection routine, like the client's `signature_algorithms` details. Not a must, but something to think about.

It looks like there is a patch open recently for OpenJDK as well http://mail.openjdk.java.net/pipermail/security-dev/2017-June/015960.html it went through several review cycles, but it's not clear if it was ever merged.

**Comment by mouse07410** [comment]
> my preferred ordering would be 1) DN 2) cert hash 3) key hash

I would probably reverse this order.

a. Public key hash conveys to the adversary nothing about the user identity (except that one can track consistent use of that key), and
b. Public key hash allows updates & modifications of the certificate without the need to change access parameters. To illustrate: some of my keys are on hardware tokens. Certificates are at a central repository. Certificate Key Usage and EKU changed a couple of times (updates in the policy and some additional use cases), but the tokens did not need to be re-issued because the keys remained valid.

**Comment by securitykernel** [comment]
Three years down the road we now really need to implement this. We only need to pass `Cert_Hash` to the server, so we would go and add only this option. In TLS 1.3, there is a different extension type `certificate_authorities` that only permits sending a DN.

Any updates to your comments from 2017 regarding implementation details?

**Comment by randombit** [comment]
Re DN vs key hash vs cert hash, I really don't have strong opinions on this. Though if 1.3 is DN-only it would be nice if we did DN-only for 1.2 as well since otherwise we can end up in some very confusing situations eg if both extensions are used (when 1.3 support is added I mean).

It would be good to know what other implementations support this extension and of those which types they support.

**Comment by securitykernel** [comment]
We discussed this internally again and since TLS 1.3 is DN-only ([link to RFC](https://tools.ietf.org/html/rfc8446#section-4.2.4)), but we need to be able to send certificate hashes in our usage scenario, we will not implement this extension. Instead, we'll use ALPN to send the certificate hashes. We are not too happy with this solution, but it is only a compatibility fix, so we can live with it.

Oh and I found no other TLS stack that implements trusted_ca_keys. My MatrixSSL finding from 2017 seems to be wrong and still is. Even the OpenJDK patch seems to not have been merged in the end.
